src/main/java/io/debezium/connector/vitess/VitessPartition.java

```diff
@@ -58,7 +63,16 @@ static class Provider implements Partition.Provider<VitessPartition> {
 
         @Override
         public Set<VitessPartition> getPartitions() {
-            return Collections.singleton(new VitessPartition(connectorConfig.getLogicalName()));
+            if (connectorConfig.offsetStoragePerTask()) {
+                String taskKey = connectorConfig.getVitessTaskKey();
+                if (taskKey == null) {
+                    throw new RuntimeException("No vitess.task.key in: {}" + connectorConfig.getConfig());
+                }
+                return Collections.singleton(new VitessPartition(taskKey));
```

Comment:
It's a bit awkward to set create VitessPartition by populating server name (i.e. connector logical name in single-task setup) with the task key.
Shall we consider enhance VitessPartition to incorporate a meaningful field for task key. Something like https://github.com/debezium/debezium-connector-vitess/pull/61/files

Reply:
I see, but #61 PR hasn't been merged in yet.  Although we are only creating one VitessPartition object for the given source task, this VitessPartition object actually are working with multiple vitess shards (same way as the old VitessParition prior to this PR).  If you don't like the "server" in the partition key names, I can call it "shards" (note it's plurals, not the singular "shard" in PR 61).

Reply:
Yes. That PR was not merged, because there was no good use cases for it. 
But now this sounds a good fit.

> If you don't like the "server" in the partition key names, I can call it "shards" (note it's plurals, not the singular "shard" in PR 61).

I meant keep the "server" (that's the logical name of the connector), and add an extra key, maybe just "task_key" since you are passing it. If you want to add "shards", you might want to pass the actual shards?

Reply:
That 'server' name as the partition key is really useless.  Maybe the initial intention was to make sure there is no key conflicts when multiple databases are going into that Kafka offset topic.  But this partition key is listed under the connector name in the kafka topic.  When there are multiple databases or keyspaces, we would have to create multiple connectors (each with distinct name).  So inside kafka offset topic, it will become {connector1, {server='db1'}} {connector2, {server='db2'}}.  So under connector1, the dbServer is always going to be db1 (even when we start to create multiple source partitions to support multiple vitess shards), that's why I don't intend to keep dbServer name as part of the source partition name.  And that dbServer name is already persisted in another kafka config: connect-configs, so that information is not lost either.

Reply:
For the offset topic, you are right, it's kind of useless.

However, the second part of the offset key actually is called "source partitions", i.e {server='db1'} part in `{connector1, {server='db1'}}`. See https://github.com/debezium/debezium/blob/main/debezium-core/src/main/java/io/debezium/pipeline/spi/Partition.java

I think it's used by Debezium event dispatcher to decide the partition key for the source record. https://github.com/debezium/debezium/blob/main/debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java#L397

That means source partition is better to be a unique one per connector and task to avoid unexpected behavior such as hot spot. it's better to keep the original semantics while incrementally adding more connector-specific semantics without breaking original ones. cc @gunnarmorling @jpechane

Reply:
This SourcePartition concept is actually not a debezium thing, it's a concept in Kafka Connect Framework: https://github.com/apache/kafka/blob/trunk/connect/api/src/main/java/org/apache/kafka/connect/source/SourceRecord.java#L33.   Kafka Connect use a Map of keys to identify a Source Partition to allow for situations that partition key is a combo key (e.g. {database=db1, table=table1}).  In the vitess connector, if we use the offset.storage.per.task mode, the partition key is based on the taskId, the database name doesn't really mean much for the source partition, the database name is not a secondary key used to further drill into the partition space.  The Map for the source offset is designed for the similar flexibility purpose.  The keys in the map for source offset doesn't need to be the same as the keys in the Map for the source partition.

Reply:
To both debezium and Kafka Connect, the SourcePartition and SourceOffset are pass-through items, it's only meaningful within the vitess connector.  As long as their use are consistent between writing and retrieving, the content of source partition is totally decided by vitess connector.  I think I can change the key name within SourcePartition to be 'taskId' to reflect the current partitioning scheme, database name is not really used in this partitioning scheme to identify anything so it doesn't really need to be included.

Reply:
I see what you meant. I don't have a strong opinion. 

My thought is that even if source partition a is passed through entity or for debug/information purpose, I still see its value if you have multiple connectors in the same Kafka Connect cluster with the same multi-task distribution, say 4 tasks, it will let the downstream components to believe that they come from the same "source partition", but in fact they are not; they come from multiple different connectors. In this case I think "make every field meaningful" makes more sense to me.

Reply:
OK, let me add a task_key param in SourcePartitions.

Reply:
@shichao-an Please take a look at today's commit: https://github.com/debezium/debezium-connector-vitess/pull/91/commits/8c3975e0ef8c012784c510f6058a6f8337ba702f which addressed this issue (adding task_key param) plus more unit/integration tests.